### PR TITLE
Improve error message 'can't qualify macro invocation with '

### PR DESCRIPTION
**Pull request by kosta**
I tried to declare a function, wrote `pub` but forgot the `fn`.

The compiler complains about something with macros.

> error: can't qualify macro invocation with `pub`
...
>  = help: try adjusting the macro to put `pub` inside the invocation
https://is.gd/VKgQLc

I added to the error message that a keyword might be missing as well.

* Is keyword the right term here?
* Is there a better way to phrase this?

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -4768,7 +4768,8 @@ impl<'a> Parser<'a> {
                 } else {
                     let mut err = self.diagnostic()
                         .struct_span_err(sp, "can't qualify macro invocation with `pub`");
-                    err.help("try adjusting the macro to put `pub` inside the invocation");
+                    err.help(concat!("try adjusting the macro to put `pub` inside the invocation.",
```
Comment (est31) [inline src/libsyntax/parse/parser.rs]: btw you can do the same without concat!, but by doing "text text text \
            text text".

Comment (kosta) [inline src/libsyntax/parse/parser.rs]: Ok, I will try this to get below 100 characters in that line!

**Comment by petrochenkov** [comment]
@kosta 
You've added the new message into `complain_if_pub_macro_diag` which is used in too many contexts. In many of them the message will be incorrect.

Ideally, the code issuing the "did you mean `pub fn`?" message should go into the end of `fn parse_item_` function (or beginning of `fn parse_macro_use_or_failure`) and implement the same logic that was implemented in [this PR](https://github.com/rust-lang/rust/pull/41282) for associated items (unfortunately that commit mixes it with lots of error recovery changes so it may not be obvious).
Basically, if the path in `pub path` has only one segment and is not followed by `!`, then you can issue the same diagnostics as `fn missing_assoc_item_kind_err` does (maybe you can even reuse that function with slight modifications).

**Comment by alexcrichton** [comment]
ping @kosta, just wanted to make sure stays on your radar! Do you have thoughts on @petrochenkov's thoughts?

**Comment by kosta** [comment]
Sorry for not responding earlier. The suggestions sound very reasonable, but I didn't have time to look at them yet. I'm extremely busy at the moment (expecting a baby), so it will probably be at least a month before I can do any progress on this issue. Sorry about that!

Is it ok to leave this hanging for a month or so? If not, feel free to close and I will re-open once I have the time do make any progress!

**Comment by Mark-Simulacrum** [comment]
No problem at all! Congratulations on the future baby; I'm going to close this for now -- please do reopen when you get a chance to make the changes though!
